Migrate Titan mailbox set up form to new mailbox list component

olaseni opened the pull request:
#### Overview 

The next set of changes 

#### Proposed Changes

* Migrates instances of the provider specific `<TitanNewMailboxList/>` to the generic `<NewMailboxList/>` for the Titan mailbox setup form. This is the form used for creating already purchased mailboxes for users; there's no need for checkout.
* For the main view involved, the JSX components are converted to their TSX counterparts
* Refactors the `useCreateTitanMailboxMutation()` hook such that the internal mutation now accepts variables in the late stage using the `mutateAsync()` function
* Installs autoFocus feature

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Have a Professional Email subscription with multiple mailboxes ready
* Checkout this branch in your local development setup, or use [Calypso-live ](https://calypso.live/?image=registry.a8c.com/calypso/app:build-35778)
* Navigate to the Email plan page for the subscription i.e. [ **Upgrades** ] > [ **Domains** ] > [ **Active domain** (i.e. example.com) ]  > [ **(\d+) mailboxes** ]
* Open another browser tab to the email provider's external control panel. You can reach that from the Email plan page, using the [ **Manage all mailboxes** ] nav item. Use any of the external nav items exposed to access the external control panel
* Remove one or more mailboxes from the subscription using the external control panel
* Back at the open tab where the Email plan page is open, confirm that the deletion change is reflected ( i.e. one less mailboxes are now displayed)
* Confirm that an `Action required` notice with a CTA to `Set up mailbox` is now visible
*  Use the CTA to reveal the new mailbox form
* Confirm that form's behaviour is identical to the former behaviour before checkout. You can establish this by opening up another tab with the same form using the production website i.e. `https://wordpress.com/***`
* Confirm that you can create a new mailbox for the subscription without going through checkout

| Illustration             |
:------------------:
<img width="600" alt="Screenshot 2022-06-18 at 4 25 09 AM" src="https://user-images.githubusercontent.com/277661/174421382-5f76d428-1e28-48a0-ae4d-68f217be8fe8.png"> |
<img width="600" alt="Screenshot 2022-06-18 at 4 25 30 AM" src="https://user-images.githubusercontent.com/277661/174421389-ad2d3181-754d-4bb0-92b8-314d11b06f4b.png"> |
<img width="600" alt="Screenshot 2022-06-18 at 4 26 17 AM" src="https://user-images.githubusercontent.com/277661/174421401-bc88a72f-6538-4287-afed-e1eddf202202.png"> |
<img width="600" alt="Screenshot 2022-06-18 at 4 26 43 AM" src="https://user-images.githubusercontent.com/277661/174421405-f7873c9b-0a27-40c7-8c05-12bce2373835.png"> |
<img width="600" alt="Screenshot 2022-06-18 at 4 38 18 AM" src="https://user-images.githubusercontent.com/277661/174421453-5694df2d-23ee-4b88-9a60-d508c3b7f94f.png"> |

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #64424

olaseni committed:
Migrate to new mailbox list

olaseni committed:
Smoothen types ...

olaseni committed:
Add autoFocus feature to enable focusing the first input element

fredrikekelund commented on client/my-sites/email/titan-set-up-mailbox/titan-set-up-mailbox-form.tsx:
```diff
@@ -0,0 +1,127 @@
+	const { canContinue, dispatch, domainName, mailboxName } = dispatchProps;
+
+	dispatch(
+		recordTracksEvent(
```
I'm curious, are there any rules of thumbs for choosing between `recordTracksEvent` from `calypso/state/analytics/actions` (Redux action creator) vs `recordTracksEvent` from `calypso/lib/analytics/tracks` (push Tracks event "directly")?

olaseni replied:
None that I know of, personally.

fredrikekelund replied:
Okay, cool! Wasn't necessarily a comment to change this code. I've seen plenty of both solutions in the codebase and I had to ask just because it seems simpler to not have to `dispatch` the event.

stephanethomas replied:
Let's ping @Automattic/team-calypso, maybe they can clarify which option is better 🙂.

tyxla replied:
Hey folks, thanks for the ping! 

We'd generally recommend **going with `lib/analytics` when possible**, because `state/analytics` is simply a wrapper around it, and we'd generally like to get rid of any unnecessary wrappers.

The rationale is simple: tracks code is spread throughout the codebase, and if we don't rely on Redux and all the underlying custom infrastructure, we have one less piece to carry around throughout the codebase.

Now, what likely makes this confusing is that there was a time when we did recommend using the Redux wrapper. However, that was in times when we worked to centralize Redux and the Calypso implementation as "one solution to rule them all". With `react-query ` and `@wordpress/data` this is no longer the case, and the less coupled we are, the better.

stephanethomas commented on client/data/emails/use-create-titan-mailbox-mutation.ts:
```diff
@@ -1,41 +1,45 @@
+};
+
+export { useCreateTitanMailboxMutation };
+export type { TitanMailboxFields };
```
We could remove those two lines by exporting the type and function directly, e.g.:

```javascript
export const useCreateTitanMailboxMutation = () => { ... }
```

It also makes it clear which element is exported looking at their declaration.

stephanethomas commented on client/my-sites/email/form/mailboxes/components/mailbox-field/index.tsx:
```diff
@@ -46,6 +48,7 @@ const MailboxFieldInput = ( {
 		onBlur,
 		onChange,
 		onInvalid,
+		...( renderPosition === 1 ? { autoFocus: true } : {} ),
```
A small suggestion but instead of passing `renderPosition` we could actually provide a boolean called `isFirstField`, which is a bit more explicit than the former.

stephanethomas commented on client/data/emails/use-create-titan-mailbox-mutation.ts:
```diff
@@ -1,41 +1,45 @@
+	[ FIELD_PASSWORD ]: string;
+	[ FIELD_ALTERNATIVE_EMAIL ]: string;
+	[ FIELD_IS_ADMIN ]: boolean;
+};
```
We could move that type declaration to the `types.ts` file (in the same folder, or even better, in `calypso/my-sites/email/form/mailboxes/constants`).

stephanethomas commented on client/my-sites/email/form/mailboxes/components/mailbox-form-wrapper/index.tsx:
```diff
@@ -27,15 +27,20 @@ const MailboxFormWrapper = ( {
 	};
 
-	const UserFormFields = () => {
+	const renderUserFormFields = () => {
```
What's the reason for renaming this? It's no longer consistent with how we named the `GoogleFormFields` and `TitanFormFields` components.

olaseni replied:
When it became important to dynamically determine the position of the form elements, I realised that inner components are essentially functions and are executed after the currently executing component has been fully evaluated.

```jsx
<>
    <div> Words and text </div>
    <DirectComponent />
    { someFunctionThatRendersComponents() }
</>
```
For example, `<DirectComponent />` will be rendered/called in its own context AFTER the `someFunctionThatRendersComponents()` is called. 

This makes determining visual element positions difficult. 
The workaround is to call a `renderer function` because then the scope of that function evaluates in the right order.

stephanethomas commented on client/my-sites/email/titan-set-up-mailbox/index.tsx:
```diff
@@ -56,10 +65,10 @@ const TitanSetUpMailbox = ( { selectedDomainName, source } ) => {
 
 			<Main wideLayout={ true }>
-				<DocumentHead title={ titleCase( title ) } />
+				<DocumentHead title={ title } />
```
Why did you remove `titleCase()`? It was here on purpose 😉.

olaseni replied:
There's no type definition for that module since it's an older one and seemingly hasn't been used in TS before. 
I figured we could achieve the same in CSS since it's such a trivial case.

However I have reconsidered and added a local type definition file that declares that module as `any`. I hope that's okay @Automattic/team-calypso ?

tyxla replied:
FWIW we can't style the title of the document with CSS - it's fixed by the browser, and specifically, the OS we're using.  

When it comes to declaring a local module - that generally works, but can we make it a bit more specific than just `any`? It seems to be a straightforward case that just accepts strings and returns strings?

olaseni replied:
> FWIW we can't style the title of the document with CSS - it's fixed by the browser, and specifically, the OS we're using

Oh, sorry about the confusion. My reference to styling was to the `<SectionHeader/>`

olaseni replied:
> but can we make it a bit more specific than just any? It seems to be a straightforward case that just accepts strings and returns strings?

Sure. Thanks for the quick feedback!

stephanethomas reviewed (approved):
I only had a few minor comments but otherwise this looks good, and works fine 👍.

![image](https://user-images.githubusercontent.com/594356/174652443-715f0574-1c46-4e67-a45a-d107be919203.png)

olaseni committed:
Implement review suggestions:
- Switch resource export style
- Swap the `renderPosition` prop on field inputs with the boolean `isFirstField`

olaseni committed:
Move a type definition to a more central file

olaseni committed:
Add a TS type definition file for `to-title-case`

olaseni commented:
> I spent some time looking at the blurring issue I described in [this comment](https://github.com/Automattic/wp-calypso/pull/64605#pullrequestreview-1007716983).
 ...  I can't see why the internal flags of the react-query hook would cause the entire DOM tree to re-render, so I'm not sure that this is a 100% satisfying solution, but still it gets to the issue at hand.

Thanks for looking into this. @fredrikekelund . I suspected this as well but was still working on a good fix. I'll add this in since it's better and works now.

olaseni committed:
Restrict render triggers to incoming data changes

olaseni committed:
Improve typings for module
